### PR TITLE
feat: Add Connector trait

### PR DESCRIPTION
**Pull request by zu1k**
I need to get a tcpStream established from a specific source IP, but I didn't find a proper `bind_connect` implementation in `ureq`.

I checked the historical issue and found a lot of discussions on related content. #301 #419 #251

At the beginning, I added the `bind_connect` implementation to `ureq`, but I quickly relized that `ureq` needs to be kept lightweight, and there may be similar requirements in the future, such as establishing connections from socks files, or other special requirements, so finally I abstracted a most basic implementation of the `Connector` trait, and implemented `BindConnector` based on this.

The definition of the `Connector` trait is the same as the `connect` and `connect_timeout` of std `TcpStream`. I think it may need to be improved according to the needs in the future, but it should be enough for now.

Signed-off-by: zu1k <i@zu1k.com>

### CODE REVIEW

**Review by mcr — COMMENTED** [review]
I am concerned about how this might interact with TLS.
Maybe the example could do both TLS and connect()?

**Comment by zu1k** [comment]
> I am concerned about how this might interact with TLS.
> Maybe the example could do both TLS and connect()?

Sorry I didn't understand you. Do you mean that the `Connector` also should do the work of the TLS handshake and stripping? Or maybe I should rename it to `TcpConnector` to avoid confusion with TlsConnector?

I don't think it's a good idea to abstract `TcpConnector` and `TlsConnector` into one `Connector`, for this makes external use of `ureq` more complicated.

File: src/connect.rs [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+/// A custom Connector to override the default TcpStream connector.
+pub trait Connector: Send + Sync {
+    fn connect(&self, addr: &SocketAddr) -> IoResult<TcpStream> {
+        TcpStream::connect(addr)
+    }
+
+    fn connect_timeout(&self, addr: &SocketAddr, timeout: Duration) -> IoResult<TcpStream> {
+        TcpStream::connect_timeout(addr, timeout)
+    }
+}
```
Comment (algesten) [inline src/connect.rs:16]: I think this should tie into the `ReadWrite` trait. https://github.com/algesten/ureq/blob/4bb6d3a4dbd7acafed801dae9d75df02207e1b87/src/stream.rs#L21

I.e. the connector trait should not produce a concrete `std::net::TcpStream`, but rather something that implements `ReadWrite`. I haven't followed through all the trait bound implications, but the starting point would be:

```rs
pub trait Connector: Send + fmt::Debug + 'static {
    type Socket;

    fn connect(&self, addr: &SocketAddr) -> io::Result<Self::Socket>;
    fn connect_timeout(&self, addr: &SocketAddr, timeout: Duration) -> io::Result<Self::Socket>;
    fn set_read_timeout(&self, socket: &mut Self::Socket);
}
```

and the agent would have something like:

```rs
    pub connector: Arc<dyn Connector<Socket = dyn ReadWrite>>,
```

Notice this abstraction would require use redo today's handling of `set_read_timeout` relying on `ReadWrite::socket()`. Unclear whether we should just go `set_read_timeout` straight on the `ReadWrite` trait.

Also minor detail, style wise we use `io::Error` instead of `IoError`.

**Review by algesten — CHANGES_REQUESTED** [review]
Overall I'm really positive to this abstraction, however, I want to tie it to the `ReadWrite` trait instead of working with `std::net::TcpStream` directly.

That potentially makes the task a lot harder, because it has wider ramifications for the handling of `set_read_timeout` as well as some potentially tricky trait bounds to work out.

File: src/connect.rs (hunk 1 above)
Comment (zu1k) [reply src/connect.rs:16]: `ureq` relies on the functions of `TcpStream` everywhere.
If we want to abstract it into a trait, we need to cover almost all of `TcpSteam`'s methods, I think this introduces too much complexity since `TcpStream` is already an abstraction for multiple platforms.

It looks like `hyper` adds a layer to be compatible with tcp-based h2 and quic-based h3, I don't think there is such a trait that can handle subsequent changes. I think abstracting away `TcpStream` should not be considered if not for the time being.

Comment (algesten) [reply src/connect.rs:16]: > If we want to abstract it into a trait, we need to cover almost all of `TcpSteam`'s methods

Really? In my mind it's pretty much only `set_read_timeout` – but maybe I'm not thinking this through?

Comment (zu1k) [reply src/connect.rs:16]: and `set_write_timeout`, `set_nonblocking`, `set_nodelay`

Comment (algesten) [reply src/connect.rs:16]: * `set_nodelay` can be set as soon as the socket has connected, no need to abstract over it (pass it as an argument to connect).
* `set_write_timeout` and `set_read_timeout` can probably be one single thing.
* `set_nonblocking` is used in a very specific way that can be abstracted.

I think we'd end up with something like:

```rs
pub trait Connector: Send + fmt::Debug + 'static {
    type Socket;

    fn connect(&self, addr: &SocketAddr, timeout: Option<Duration>, nodelay: bool) -> io::Result<Self::Socket>;
    fn set_read_write_timeout(&self, socket: &mut Self::Socket) -> io::Result<()>;
    fn check_server_closed(&self, socket: &mut Self::Socket) -> io::Result<()>;
}
```

Sure. It's a bigger change working this into the entire code base. Before embarking on it, maybe we should broaden the discussion.

@jsha what are your thoughts on this?

Comment (zu1k) [reply src/connect.rs:16]: Agree, this requires more discussion.

**Comment by amunra** [comment]
I've been redirected here from #692. My requirement is pretty simple: Set the local bind interface address before establishing the connection, something that [`std::net::TcpStream`](https://doc.rust-lang.org/std/net/struct.TcpStream.html) can't do, thus it would be nice to allow the `ureq` user to do this by setting a connector on the agent.

It's probably a good idea to start summarising everything and document it in one go before jumping to a propoal.

Reading through the comments above there seem to be two candidate approaches with different tradeoffs and a set of desired features.

I'll start with the desired features (aka constraints):
* `(C1)` No breaking API changes.
* `(C2)` The ability to construct a custom `TcpStream`, for example with `.bind()` calls or custom TCP flags.
* `(C3)` Is composable with `ureq::stream::TlsConnector`. 
* `(C4)` Is compatible with `ureq::stream::Stream`:
    * `Stream` exposes `pub(crate) remote_addr: SocketAddr,`
* `(C5)` Is compatible with `ureq::stream::ReadWrite`:
    * `ReadWrite` exposes `fn socket(&self) -> Option<&TcpStream>`
* `(C6)` The ability to more generically connect a URL to something that can read and write, thus supporting other protocols such as unix domain sockets.

So far, there are two proposed approaches with partial coverage of these constraints:
* `(A1)` support a new tcp-specific connection trait that yields a connected `TcpStream`.
    * Satisfies `(C1)`, `(C2)`, `(C3)`, `(C4)`, `(C5)`.
* `(A2)` support a new generic connection trait that returns a `dyn ureq::stream::ReadWrite`.
    * Satisfies `(C1)`, `(C2)`, `(C4)`, `(C6)`.
    * Incompatible with `(C3)` as there may not be a DNS available to perform TLS auth.
    * Possibly incompatible with `(C5)`? **[Feedback desired]**
        *  Unclear if can fully work with `ReadWrite`, since all examples I could see return `Some(socket)` from `fn socket`: I don't fully understand if it's OK for it this field to be `None`.

TL;DR of the two approaches:
* `(A1)`:
    * Is simple to implement and slots right into the existing code.
    * Isn't generic enough for `(C6)`.
    * Easier to use for most users who mostly care about TCP.
    * Easily composes with TLS.
* `(A2)`:
    * Is trickier to implement due to the rest of the code relying on socket.
    * Is harder to use for users just caring about sockets (but examples can help here).
    * Is not (necessarily) composable with TLS, unless a DNS can be exposed.

@algesten (and others), kindly let me know where I'm going wrong and what my misunderstandings are: I've only stared at the code for 20 minutes :-).

**Comment by amunra** [comment]
Can you point me to the timeout code you're referring to?

**Comment by amunra** [comment]
I've taken a quick look at what would be involved in `(A2)`.
The first step would be to decouple `struct Stream` from `std::net::TcpStream`.
https://github.com/algesten/ureq/compare/main...amunra:ureq:connector

This is enough to compile the main code, but not the tests and doesn't even introduce a connnector trait yet.

Personally, I feel like sticking to TCP (i.e. implementing `(A1)`) may be a more pragmatic apprach, at the cost of dropping `(C6)`.

I think `(A2)` _can_ be brought to completion, but would complicate the existing logic a fair bit: The code I hacked so far feels very much square-peg-round-hole and cleaning (changing the `ReadWrite` trait) would require breaking API compatibility `(C1)`.

@algesten I'd appreciate your thoughts.

**Comment by algesten** [comment]
Fair enough. Let's go with the pragmatic choice, A1.

**Comment by amunra** [comment]
@zu1k, I wonder if it would make sensible to rename the trait to `TcpConnector`, this would allow more scope for a more general non-tcp specific connector to be written afterwards, should there be renewed scope for it.

**Comment by zu1k** [comment]
> @zu1k, I wonder if it would make sensible to rename the trait to `TcpConnector`, this would allow more scope for a more general non-tcp specific connector to be written afterwards, should there be renewed scope for it.

It should be so.

**Comment by algesten** [comment]
A variant of this is done in 3.x
